### PR TITLE
ArrayDataAdapter no longer converts to NumPy and supports sparse tens…

### PR DESCRIPTION
**Pull request by hertschuh**
…ors.

Instead, the passed arrays can be sliced or indexed in their native format.
- This addresses https://github.com/keras-team/keras/issues/18408 and improves performance, especially with Tensorflow and Torch. It improves TF -> TF and Torch -> Torch, but also TF -> Torch and Torch -> TF.
- This allows the support of sparse tensors (`tf.SparseTensor`, `jax.experimental.sparse.BCOO` and `scipy.sparse`). These sparse tensors are sliced as sparse and the iterators yield sparse tensors in the requested format (either TF or JAX).
- The `validation_split` argument of `Model.fit()` can now be used with anything supported by `ArrayDataAdapter`, in particular, sparse tensors are now supported.

In summary, `ArrayDataAdapter` now supports:
- native Python arrays
- NumPy arrays
- Tensorflow tensors, ragged tensors, sparse tensors (new)
- JAX arrays and BCOO sparse tensors (new)
- pandas DataFrames
- pandas Series
- scipy sparse matrices (new)

Also:
- Fixed bug where batch level shuffling would shuffle inconsistently the different arrays (in particular inputs and labels) when using a TF dataset or a NumPy iterator.
- Fixed bug where `tf.RaggedTensor`s would only work when using a TF dataset.
- Fixed bug where `tf.RaggedTensor`s would not work when doing batch level shuffling.
- Added a workaround for a bug where `tf.cast`ing a `tf.SparseTensor` would lose the static shape.
- Added test coverage for `tf.RaggedTensor`s and `pandas.Series`.
- Added verification in tests that inputs and labels are shuffled consistently.

### CODE REVIEW

**Comment by hertschuh** [comment]
> Great work! Are you seeing any performance impact?

Yes, here is the summary of the benchmark. I used a trivial model, clocked `model.fit()`, this is the time difference as a percentage, so a negative value is good (-50% means 2x faster).

Summary:
- Torch to Torch is a lot faster
- TF to TF is quite a bit faster
- the rest is about the same

|  ↓ Input ↓ \ Backend (output) → | Tensorflow | JAX | Torch |
|---------|--------|--------|--------|
| Numpy arrays          | 0 to -21%         | -3% to -4% | -1% to -2% |
| Tensorflow tensors | -23% to -44% | 14% to -1%‡ | -18% |
| JAX arrays               | 0 to -20%         | -3% to -9% | -2% |
| Torch tensors          | -14% to -32%  | 2% to -14% | -53% to -55% |

Notes:
- it's a range not a single value because performance impact is different for `shuffle=True / False / "batch"`
- Anything less than 4% is not significant
- ‡ TF tensors to JAX is a bit slower only when `shuffle=False`

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
